### PR TITLE
Add `OwnershipAssignment` to dbt `VirtualView`

### PR DESCRIPTION
**Pull request by usefulalgorithm**
### 🤔 Why?

Enables assigning ownership to dbt model's virtual views, or both the virtual view and the dataset itself.

Tracked by [SC20587](https://app.shortcut.com/metaphor-data/story/20587)

### 🤓 What?

- Added a config field for meta_ownerships:
```yaml
meta_ownerships:
  - meta_key: owner
    ownership_type: Data Steward
    assignment_target: materialized_table # assigns the owner to either "materialized_table" (the dataset), "dbt_model" (the virtual view), or "both". Default is the materialized table.
```
- Added support for parsing string values in dbt model's meta field, such that a string value containing commas will be parsed as a sequence of string values

### 🧪 Tested?

Tested with the newly added config, the ownership information was added correctly.

### CODE REVIEW

**Review by mars-lan — COMMENTED** [review]
Let's also add a config to control the desired behavior, i.e., assigning owners to dbt model, materialized tables, or both.

File: metaphor/dbt/util.py [hunk 1]
```diff
@@ -71,24 +78,31 @@ def to_owner(email_or_username: str, email_domain: Optional[str]) -> Optional[st
...
         return str(to_person_entity_id(email))
 
-    ownerships: List[Ownership] = []
+    ownerships = Ownerships()
     for meta_ownership in meta_ownerships:
         value = meta.get(meta_ownership.meta_key)
         email_or_usernames = []
         if isinstance(value, str):
-            email_or_usernames.append(value)
+            # If there's a comma in the string, treat it as a sequence of strings
```
Comment (usefulalgorithm) [inline metaphor/dbt/util.py:86]: I saw a model with owner = 'mars@metaphor.io,pardhu@metaphor.io', and that's definitely not a single user email!

**Review by elic-eon — DISMISSED** [review]
LGTM!

File: metaphor/dbt/README.md [hunk 2]
```diff
@@ -72,6 +72,15 @@ meta_ownerships:
     email_domain: test.com
 ```
 
+You can also choose between assigning the owner to the materialized table, the dbt model, or both:
+
+```yaml
+meta_ownerships:
+  - meta_key: owner
+    ownership_type: Data Steward
+    assignment_target: dbt_model # Valid choices: "dbt_model", "materialized_table", both. Default is "materialized_table"
```
Comment (mars-lan) [inline metaphor/dbt/README.md:81]:
```suggestion
    assignment_target: dbt_model # Valid choices: "dbt_model", "materialized_table", "both". Default is "materialized_table"
```

File: metaphor/dbt/config.py [hunk 3]
```diff
@@ -1,11 +1,13 @@
 from dataclasses import field as dataclass_field
-from typing import List, Optional
+from typing import List, Literal, Optional
 
 from pydantic.dataclasses import dataclass
 
 from metaphor.common.base_config import BaseConfig
 from metaphor.common.dataclass import ConnectorConfig
 
+MetaOwnershipAssignmentTarget = Literal["dbt_model", "materialized_table", "both"]
```
Comment (mars-lan) [inline metaphor/dbt/config.py:9]: Is this the preferred method for Pydantic 2.x? We've been using enums before I believe?

File: metaphor/dbt/README.md (hunk 2 above)
Comment (usefulalgorithm) [reply metaphor/dbt/README.md:81]: ah missed it

File: metaphor/dbt/util.py (hunk 1 above)
Comment (mars-lan) [reply metaphor/dbt/util.py:86]: I thought the YAML parser would automatically turn comma-separated strings into lists?

File: metaphor/dbt/config.py (hunk 3 above)
Comment (usefulalgorithm) [reply metaphor/dbt/config.py:9]: yes we have been using enums but I believe this is also fine. If you specify something illegal it just breaks, for example:

```yaml
meta_ownerships:
  - meta_key: owner
    ownership_type: Data Steward
    assignment_target: gg
```

the result:
```
2023-11-01 17:17:01:INFO:metaphor:Executing dbt.cloud connector with config file ../staging_config/DbtCloud.763c0e54-5d03-47e0-84b7-d42523b5835a.yml
Traceback (most recent call last):
  File "/Users/andy/Library/Caches/pypoetry/virtualenvs/metaphor-connectors-np3LZT1s-py3.9/bin/metaphor", line 6, in <module>
    sys.exit(main())
  File "/Users/andy/work/connectors/metaphor/__main__.py", line 35, in main
    package_main(args.config)
  File "/Users/andy/work/connectors/metaphor/dbt/cloud/__init__.py", line 6, in main
    cli_main(DbtCloudExtractor, config_file)
  File "/Users/andy/work/connectors/metaphor/common/cli.py", line 7, in cli_main
    extractor = extractor_cls.from_config_file(config_file)
  File "/Users/andy/work/connectors/metaphor/dbt/cloud/extractor.py", line 22, in from_config_file
    return DbtCloudExtractor(DbtCloudConfig.from_yaml_file(config_file))
  File "/Users/andy/work/connectors/metaphor/common/base_config.py", line 36, in from_yaml_file
    return TypeAdapter(cls).validate_python(variable_substitution(obj))
  File "/Users/andy/Library/Caches/pypoetry/virtualenvs/metaphor-connectors-np3LZT1s-py3.9/lib/python3.9/site-packages/pydantic/type_adapter.py", line 283, in validate_python
    return self.validator.validate_python(__object, strict=strict, from_attributes=from_attributes, context=context)
pydantic_core._pydantic_core.ValidationError: 1 validation error for DbtCloudConfig
meta_ownerships.0.assignment_target
  Input should be 'dbt_model', 'materialized_table' or 'both' [type=literal_error, input_value='gg', input_type=str]
    For further information visit https://errors.pydantic.dev/2.4/v/literal_error
```

File: metaphor/dbt/util.py (hunk 1 above)
Comment (usefulalgorithm) [reply metaphor/dbt/util.py:86]: maybe the yaml parser does that, but `value` here does not come from the yaml config - it is from the meta field in the dbt model

Comment (mars-lan) [reply metaphor/dbt/util.py:86]: Got it. Let's update https://github.com/MetaphorData/connectors/blob/main/metaphor/dbt/README.md to include that behavior.

Comment (mars-lan) [reply metaphor/dbt/util.py:86]: Actually no, we should change the incorrect example (`mars@metaphor.io,pardhu@metaphor.io`) into a list and update the doc.

**Review by elic-eon — DISMISSED** [review]

**Review by mars-lan — CHANGES_REQUESTED** [review]
Let's address https://github.com/MetaphorData/connectors/pull/653#discussion_r1378576231 before merging.

File: metaphor/dbt/util.py (hunk 1 above)
Comment (usefulalgorithm) [reply metaphor/dbt/util.py]: modified the incorrect example in another repo, and reverted this change.

File: metaphor/dbt/config.py [hunk 4]
```diff
@@ -18,6 +20,10 @@ class MetaOwnership:
     # Domain for user names
     email_domain: Optional[str] = None
 
+    # The target to assign this ownership to. Can be either the dbt model, the materialized table, or both.
+    # Defaults to the materialized table.
+    assignment_target: MetaOwnershipAssignmentTarget = "materialized_table"
```
Comment (mars-lan) [inline metaphor/dbt/config.py:25]: Default to `both` as that's most likely the expected behavior.

File: metaphor/dbt/util.py [hunk 5]
```diff
@@ -83,12 +90,14 @@ def to_owner(email_or_username: str, email_domain: Optional[str]) -> Optional[st
...
-                ownerships.append(
-                    Ownership(
-                        contact_designation_name=meta_ownership.ownership_type,
-                        person=owner,
-                    )
+                ownership = Ownership(
+                    contact_designation_name=meta_ownership.ownership_type,
+                    person=owner,
                 )
+                if meta_ownership.assignment_target != "materialized_table":
```
Comment (mars-lan) [inline metaphor/dbt/util.py:97]: While this logic works, it's less readable than 

```
if meta_ownership.assignment_target in ['both', 'dbt_model']:
```

**Review by mars-lan — COMMENTED** [review]
Almost there!

**Review by mars-lan — APPROVED** [review]
Great work!

**Pull request merged**